app/services/git_hub_authenticator.rb

```diff
@@ -0,0 +1,57 @@
+class GitHubAuthenticator
+  def initialize(auth, current_user = nil)
+    @auth = auth
+    @current_user = current_user
+  end
+
+  def self.authenticate(auth, current_user = nil)
+    new(auth, current_user).authenticate
+  end
+
+  def authenticate
+    if user = find_user
+      user.update_attributes(github_params)
+      user
+    else
+      User.create(user_params)
+    end
+  end
+
+  private
+
+  attr_reader :auth, :current_user
+
+  def find_user
+    @user ||= current_user ||
+      User.find_by(github: auth.info.nickname) ||
+      User.find_by(email:  auth.info.email)
+  end
+
+  def auth_email
+    @auth_email ||= auth.info.email
+  end
+
+  def token
+    @token ||= auth.credentials.token
+  end
+
+  def github_email
+    return auth_email if auth_email.present?
+    GitHubBub.get("/user/emails", token: token).json_body.first
+  end
+
+  def github_params
+    @github_params ||= {
```

Comment:
It's also called from `user_params` as well. `user_params` use that as foundation